File: backend/src/models/Budget.js
const mongoose = require('mongoose');
const crypto = require('crypto');

const BudgetSchema = new mongoose.Schema(
  {
    budgetId: {
      type: String,
      default: () => crypto.randomUUID(),
      unique: true,
      index: true
    },
    userId: {
      type: String,
      ref: 'User',
      required: [true, 'User ID is required'],
      index: true
    },
    category: {
      type: String,
      required: [true, 'Category is required'],
      enum: {
        values: [
          'Food', 'Transport', 'Shopping', 'Bills', 'Entertainment',
          'Healthcare', 'Education', 'Rent', 'Utilities', 'Others'
        ],
        message: '{VALUE} is not a valid budget category'
      }
    },
    amount: {
      type: Number,
      required: [true, 'Budget amount is required'],
      min: [0, 'Amount must be greater than or equal to 0']
    },
    month: {
      type: Number,
      required: [true, 'Month is required'],
      min: [1, 'Month must be between 1 and 12'],
      max: [12, 'Month must be between 1 and 12']
    },
    year: {
      type: Number,
      required: [true, 'Year is required']
    },
    spent: {
      type: Number,
      default: 0,
      min: [0, 'Spent amount must be non-negative']
    }
  },
  {
    timestamps: true,
    toJSON: {
      transform: function (doc, ret) {
        delete ret._id;
        delete ret.__v;
        return ret;
      }
    }
  }
);

// Compound unique index so a user cannot have two budget targets for the exact same category/month/year.
BudgetSchema.index({ userId: 1, category: 1, month: 1, year: 1 }, { unique: true });

module.exports = mongoose.model('Budget', BudgetSchema);

File: flutter_app/lib/models/budget.dart
class BudgetModel {
  final String budgetId;
  final String userId;
  final String category;
  final double amount;
  final int month;
  final int year;
  final double spent;

  BudgetModel({
    required this.budgetId,
    required this.userId,
    required this.category,
    required this.amount,
    required this.month,
    required this.year,
    required this.spent,
  });

  factory BudgetModel.fromJson(Map<String, dynamic> json) {
    return BudgetModel(
      budgetId: json['budgetId'] ?? '',
      userId: json['userId'] ?? '',
      category: json['category'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      month: json['month'] ?? 1,
      year: json['year'] ?? DateTime.now().year,
      spent: (json['spent'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'budgetId': budgetId,
      'userId': userId,
      'category': category,
      'amount': amount,
      'month': month,
      'year': year,
      'spent': spent,
    };
  }
}

class BudgetProgressModel {
  final String budgetId;
  final String category;
  final double amount;
  final double spent;
  final double remaining;
  final double percentage;
  final bool isOverBudget;

  BudgetProgressModel({
    required this.budgetId,
    required this.category,
    required this.amount,
    required this.spent,
    required this.remaining,
    required this.percentage,
    required this.isOverBudget,
  });

  factory BudgetProgressModel.fromJson(Map<String, dynamic> json) {
    return BudgetProgressModel(
      budgetId: json['budgetId'] ?? '',
      category: json['category'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      spent: (json['spent'] ?? 0).toDouble(),
      remaining: (json['remaining'] ?? 0).toDouble(),
      percentage: (json['percentage'] ?? 0).toDouble(),
      isOverBudget: json['isOverBudget'] ?? false,
    );
  }
}
